File: algorithms/medium/maximum_difference_between_node_and_ancestor.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec max_ancestor_diff(root :: TreeNode.t() | nil) :: integer
  def max_ancestor_diff(root) do
    aux(root, root.val, root.val)
  end

  def aux(nil, mi, ma), do: ma - mi

  def aux(root, mi, ma) do
    mi = min(mi, root.val)
    ma = max(ma, root.val)
    left = aux(root.left, mi, ma)
    right = aux(root.right, mi, ma)
    max(left, right)
  end
end

File: algorithms/medium/maximum_difference_between_node_and_ancestor.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

fn aux(root: &Option<Rc<RefCell<TreeNode>>>, min: i32, max: i32) -> i32 {
    if root.is_none() {
        return max - min;
    }

    let node = root.as_ref().unwrap().borrow();
    let min = min.min(node.val);
    let max = max.max(node.val);
    let left = aux(&node.left, min, max);
    let right = aux(&node.right, min, max);
    left.max(right)
}

impl Solution {
    pub fn max_ancestor_diff(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        aux(&root, i32::MAX, i32::MIN)
    }
}
